public/app/features/scenes/core/SceneObjectBase.tsx

```diff
@@ -52,69 +59,105 @@ export abstract class SceneObjectBase<TState extends SceneObjectState = {}> impl
   }
 
   private setParent() {
-    for (const propValue of Object.values(this.state)) {
-      if (isSceneObject(propValue)) {
-        propValue.parent = this;
+    for (const propValue of Object.values(this._state)) {
+      if (propValue instanceof SceneObjectBase) {
+        propValue._parent = this;
       }
 
       if (Array.isArray(propValue)) {
         for (const child of propValue) {
-          if (isSceneObject(child)) {
-            child.parent = this;
+          if (child instanceof SceneObjectBase) {
+            child._parent = this;
           }
         }
       }
     }
   }
 
-  /** This function implements the Subscribable<TState> interface */
-  subscribe(observer: Partial<Observer<TState>>) {
-    return this.subject.subscribe(observer);
+  /**
+   * Subscribe to the scene state subject
+   **/
+  subscribeToState(observerOrNext?: Partial<Observer<TState>>): Subscription {
+    return this._subject.subscribe(observerOrNext);
+  }
+
+  /**
+   * Subscribe to the scene event
+   **/
+  subscribeToEvent<T extends BusEvent>(eventType: BusEventType<T>, handler: BusEventHandler<T>): Unsubscribable {
+    return this.events.subscribe(eventType, handler);
   }
 
   setState(update: Partial<TState>) {
-    const prevState = this.state;
-    this.state = {
-      ...this.state,
+    const prevState = this._state;
+    this._state = {
+      ...this._state,
       ...update,
     };
     this.setParent();
-    this.subject.next(this.state);
+    this._subject.next(this._state);
 
-    // broadcast state change. This is event is subscribed to by UrlSyncManager and UndoManager
-    this.getRoot().events.publish(
+    // Bubble state change event. This is event is subscribed to by UrlSyncManager and UndoManager
+    this.publishEvent(
       new SceneObjectStateChangedEvent({
         prevState,
-        newState: this.state,
+        newState: this._state,
         partialUpdate: update,
         changedObject: this,
-      })
+      }),
+      true
     );
   }
 
-  private getRoot(): SceneObject {
-    return !this.parent ? this : this.parent.getRoot();
+  /*
+   * Publish an event and optionally bubble it up the scene
+   **/
+  publishEvent(event: BusEvent, bubble?: boolean) {
+    this.events.publish(event);
+
+    if (bubble && this.parent) {
+      this.parent.publishEvent(event);
+    }
```

Comment:
Should we pass `bubble` to `parent.publishEvent` as well to make it actually bubble all the way up?

Reply:
Yes!! 🤦😃